ONNX Runtime benchmark (#18)
zmaslova/dl-benchmark: src/onnxruntime_benchmark/README.md

# ONNX Runtime Benchmark
The tool allows to measure deep learning models inference performance with [ONNX Runtime](https://github.com/microsoft/onnxruntime). This implementation uses [OpenVINO Benchmark C++ tool](https://github.com/openvinotoolkit/openvino/tree/master/samples/cpp/benchmark_app) as a reference and stick to its testing methodology, thus performance results comparison of both tools is valid.
## Prerequisites
The tool was tested on Ubuntu 20.04 (64-bit) with default GCC* 9.4.0
1. CMake 3.13 or higher
2. GCC 9.4 or higher

## Build ONNX Runtime
1. Clone repository, checkout to the latest stable release and update submodules:
```
git clone  https://github.com/microsoft/onnxruntime.git
cd onnxruntime
git checkout v1.12.1
git submodule update --init --recursive
```
2. Configure it with `cmake`:

> ```

Reviewer: `mkdir build && cd build` is missing?

Author: Да, упустил

> cmake -DCMAKE_INSTALL_PREFIX=install -DCMAKE_BUILD_TYPE=Release -Donnxruntime_BUILD_FOR_NATIVE_MACHINE=ON -Donnxruntime_BUILD_UNIT_TESTS=OFF -Donnxruntime_USE_AUTOML=OFF -Donnxruntime_BUILD_SHARED_LIB=ON -Donnxruntime_USE_FULL_PROTOBUF=ON -Donnxruntime_USE_OPENMP=ON ../cmake

Reviewer: CMake Warning:
  Manually-specified variables were not used by the project:

    onnxruntime_USE_AUTOML
    onnxruntime_USE_OPENMP
    
  Looks like these variables is out-of-date, can be removed

Author: Question marked as resolved, do I still need to remove it?

Reviewer: nevermind, it's very minor comment

Author: Removed

# ONNX Runtime Benchmark

> The tool allows to measure deep learning models inference performance with [ONNX Runtime](https://github.com/microsoft/onnxruntime). This implementation uses [OpenVINO Benchmark C++ tool](https://github.com/openvinotoolkit/openvino/tree/master/samples/cpp/benchmark_app) as a reference and stick to its testing methodology, thus performance results comparison of both tools is valid.

Reviewer 2: ```suggestion
The tool allows to measure deep learning models inference performance with [ONNX Runtime](https://github.com/microsoft/onnxruntime). This implementation inspired by [OpenVINO Benchmark C++ tool](https://github.com/openvinotoolkit/openvino/tree/master/samples/cpp/benchmark_app) as a reference and stick to its measurement methodology, thus provide consistent performance results.
```

Reviewer 2: try to polish some wording

                                                     Ex.: [123.675,116.28,103.53] or with specifying inputs src[255,255,255]
        [-scale <R G B>]                             Scale values per channel for input image.
                                                     Applicable only for models with image inputs.
                                                     Ex.: [58.395,57.12,57.375] or with specifying inputs src[255,255,255]
        [-nthreads <NUMBER>]                         number of threads to utilize.
        [-nireq <NUMBER>]                            number of inference requests. If not provided, default value is set.
        [-niter <NUMBER>]                            number of iterations. If not provided, default time limit is set.
        [-t <NUMBER>]                                time limit for inference in seconds
        [-save_report]                               save report in JSON format.
        [-report_path <PATH>]                        destination path for report.
```

### Basic usage
To run tool with default options you should provide only path to model file in `.onnx` format
```

> ./benchmark_app -m model.onnx

Reviewer 3: ```suggestion
./onnxruntime_benchmark -m model.onnx
```

```
By default, the application makes inference on randomly-generated tensors for 60 seconds. For inference only default provider (`CPU` device) is available for now.

### Inputs
To pass inputs for model use `-i` option.
```
./onnxruntime_benchmark -m model.onnx -i <path_to_input_data>
```
By default, number of inputs to use determined based on model's batch size. If yout want to make inference on some files set, you can specify the number of files to take with `-nireq` option.

If the model has several inputs, files or folders must be specified for each:
```
./onnxruntime_benchmark -m model.onnx -i input1:file1 input2:file2 input3:file3
```

### Shape and layout options
To make inference with dynamic model, you must specify shape for every model's input:
```
./onnxruntime_benchmark -m model.onnx -i input1:file1 input2:file2 input3:file3 -shape input1[N,C],input1[N,C],input1[N,C]
```
Or if all inputs have the same shape you could pass just `-shape [N,C]`. The same rules are applied to `-layout` option.

### Report
To save a report with tool configuration and performance results specify `-save_report` flag:
```

> ./benchmark_app -m model.onnx -save_report -report_path report/report.json

Reviewer 3: ```suggestion
./onnxruntime_benchmark -m model.onnx -save_report -report_path report/report.json
```

[ INFO ] Device: CPU
[ INFO ]        Threads number: DEFAULT
[Step 4/8] Configuring input of the model
[ WARNING ] Layout will be detected automatically, as it wasn't provided explicitly.
[ INFO ] Set batch to 1
[Step 5/8] Setting execution parameters
[ INFO ] Default time limit is set: 60 seconds
[Step 6/8] Creating input tensors
[ INFO ] Input config 0
[ INFO ]        data (NCHW FP32 [1,3,224,224])
[ INFO ]                ../../../test_data/ILSVRC2012_val_00000001.JPEG
[ INFO ] Input config 1
[ INFO ]        data (NCHW FP32 [1,3,224,224])
[ INFO ]                ../../../test_data/ILSVRC2012_val_00000002.JPEG
[ INFO ] Input config 2
[ INFO ]        data (NCHW FP32 [1,3,224,224])
[ INFO ]                ../../../test_data/ILSVRC2012_val_00000003.JPEG
[Step 7/8] Measuring model performance (3 inference requests, limits: 60000 ms)
[ INFO ] Warming up inference took 45.46 ms
[Step 8/8] Saving statistics report
[ INFO ] Count: 1862 iterations
[ INFO ] Duration: 60002.58 ms
[ INFO ] Latency:
[ INFO ]        Median   34.41 ms
[ INFO ]        Average: 32.22 ms
[ INFO ]        Min:     19.70 ms
[ INFO ]        Max:     84.00 ms
[ INFO ] Throughput: 29.06 FPS
```

> 

Reviewer 3: ```suggestion
```

> 

Reviewer 3: ```suggestion
```
